---
title: "Rarfraction summary"
author: "Oded Sabah"
date: "`r Sys.Date()`"
output: html_document
editor_options: 
  chunk_output_type: console
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

```{r}
library(tableone)
library(kableExtra)
library(dplyr)
library(ggplot2)
library(patchwork)
library(tidyr)
```

```{r}
### table 3
metagenomic_samples <- data.frame(
  sample_ID = c("Sample", "SAMEA110452918", "SAMEA110452924", "SAMN00040286", "SAMN00036649", "SAMN00037108", "SAMN00146983", "SAMN00099612", "SAMN00143093"),
  Sequencing_technology = c("Sequencing technology", "NovaSeq 6000, 2x150 bp", "NovaSeq 6000, 2x150 bp", "Genome Analyzer II, 2x100", "Genome Analyzer II, 2x100", "Genome Analyzer II, 2x100", "HiSeq 2000, 2x100", "HiSeq 2000, 2x100", "HiSeq 2000, 2x100"),
  cohort_studies = c("Cohort", "Federici et al., 2022", "Federici et al., 2022", "HMP1", "HMP1", "HMP1", "HMP2", "HMP2", "HMP2"),
  number_of_reads = c("Number of reads (M)", "258.36", "189.18", "149.76", "149.16", "146.07", "131.49", "116.72", "112.29"),
  sequencing_depth = c("Sequencing depth (Gbp)", "38.72", "28.44", "14.98", "14.90", "14.60", "13.50", "11.67", "11.23"),
  Number_of_species = c("Number of species", "229", "357", "403", "394", "139", "221", "197", "186")
)

# Separate the header row and use it as column names
headers <- as.character(metagenomic_samples[1, ])
metagenomic_samples <- metagenomic_samples[-1, ]
colnames(metagenomic_samples) <- headers

# Reset row names
rownames(metagenomic_samples) <- NULL
```

```{r}
 tab <- kable(metagenomic_samples, format = "latex", booktabs = TRUE) %>%
  kable_styling(latex_options = c("striped", "scale_down"))%>% 
  row_spec(2, extra_latex_after = "\\midrule") %>% 
  row_spec(5, extra_latex_after = "\\midrule") %>% 
   footnote(general = c("(M) - Million", "(Gbp) - Giga base pairs"),
           general_title = "Note: ")

  
save_kable(tab, file = "~/metaanalysis/Rarefaction/Code/CALCULATE_THRESHOLD/Rarefaction_samples/simulation_ID_Table1.pdf")
```

```{r}
### table 4
alpha_summray <- data.frame(
  Measure = c("Chao1 (richness)", "Simpson (diversity)", "Shannon (evenness)", "Chao1 (richness) ", "Simpson (diversity) ", "Shannon (evenness) "),
  Participant_A_Median_IQR = c("87.5 [82.5, 92]", "0.95 [0.95, 0.96]", "5.3 [5.11, 5.4]", "130 [110.5, 133.25]", "0.96 [0.96, 0.97]", "5.88 [5.7, 6.04]"),
  Participant_A_Max_difference = c(30, 0.9, 1.36, 48, 0.04, 1.28),
  Participant_B_Median_IQR = c("66 [64, 70]", "0.94 [0.92, 0.94]", "4.8 [4.64, 4.92]", "98 [92, 103]", "0.95 [0.93, 0.96]", "5.33 [5.1, 5.5]"),
  Participant_B_Max_difference = c(37, 0.6, 1.17, 47, 0.04, 1.28)
)

colnames(alpha_summray) <- c(" ", "Median, [IQR]", "Max difference", "Median, [IQR]", "Max difference")
rownames(alpha_summray) <- alpha_summray$` `
alpha_summray$` ` <- NULL 
```

```{r}
 tab <- kable(alpha_summray, format = "latex", booktabs = TRUE) %>%
  kable_styling(latex_options = c("striped", "scale_down")) %>% 
  pack_rows("Genus level", which(rownames(alpha_summray) == "Chao1 (richness)"), which(rownames(alpha_summray) == "Shannon (evenness)"), extra_latex_after = "\\midrule") %>%
  row_spec(which(rownames(alpha_summray) == "Shannon (evenness)"), extra_latex_after = "\\midrule") %>% 
  pack_rows("Species level", which(rownames(alpha_summray) == "Chao1 (richness) "), which(rownames(alpha_summray) == "Shannon (evenness) "), extra_latex_after = "\\midrule") %>%
  row_spec(which(rownames(alpha_summray) == "Shannon (evenness) "), extra_latex_after = "\\midrule") %>% 
  add_header_above(c("","Participant A" = 2, "Participant B"= 2))  %>% column_spec(c(1,3,5), border_right = TRUE)


  
save_kable(tab, file = "~/metaanalysis/Rarefaction/Code/CALCULATE_THRESHOLD/Rarefaction_samples/alpha_summray.pdf")
```

```{r}
### table 5
beta_summray <- data.frame(
  Measure = c("Jaccard (Similarity)", "Bray-Curtis (Dissimilarity)","Jaccard (Similarity) ", "Bray-Curtis (Dissimilarity) "),
  Participant_A_Median_IQR = c("0.65 [0.56, 0.74]", "0.36 [0.22, 0.46]", "0.61 [0.5, 0.71]", "0.38 [0.25, 0.5]"),
  Participant_A_Max_difference = c(0.65, 0.71, 0.72, 0.72),
  Participant_B_Median_IQR = c("0.66 [0.55, 0.72]", "0.37 [0.25, 0.52]", "0.59 [0.48, 0.7]", "0.39 [0.27, 0.54]"),
  Participant_B_Max_difference = c(0.55, 0.68, 0.58, 0.71)
)

colnames(beta_summray) <- c(" ", "Median, [IQR]", "Max difference", "Median, [IQR]", "Max difference")
rownames(beta_summray) <- beta_summray$` `
beta_summray$` ` <- NULL 
```

```{r}
 tab <- kable(beta_summray, format = "latex", booktabs = TRUE) %>%
  kable_styling(latex_options = c("striped", "scale_down")) %>% 
  pack_rows("Genus level", which(rownames(beta_summray) == "Jaccard (Similarity)"), which(rownames(beta_summray) == "Bray-Curtis (Dissimilarity)"), extra_latex_after = "\\midrule") %>%
  row_spec(which(rownames(beta_summray) == "Bray-Curtis (Dissimilarity)"), extra_latex_after = "\\midrule") %>% 
  pack_rows("Species level", which(rownames(beta_summray) == "Jaccard (Similarity) "), which(rownames(beta_summray) == "Bray-Curtis (Dissimilarity) "), extra_latex_after = "\\midrule") %>% row_spec(which(rownames(beta_summray) == "Bray-Curtis (Dissimilarity) "), extra_latex_after = "\\midrule") %>% 
  add_header_above(c("","Participant A" = 2, "Participant B"= 2))  %>% column_spec(c(1,3,5), border_right = TRUE)


  
save_kable(tab, file = "~/metaanalysis/Rarefaction/Code/CALCULATE_THRESHOLD/Rarefaction_samples/beta_summray.pdf")
```

```{r}
# Read the CSV file
species_abundance <- read.csv("~/Downloads/Eran_Elinav_species_abundance.csv")

# Define thresholds
thresholds <- c(0, 0.001, 0.01, 0.1, 1)

# Function to apply threshold and filter rows
apply_threshold <- function(df, threshold, col) {
  df[[col]] <- ifelse(df[[col]] < threshold, 0, df[[col]])
  df_filtered <- df[df[[col]] != 0, ]
  df_filtered <- na.omit(df_filtered)
  df_filtered <- df_filtered[!grepl("^t__", df_filtered$Abundance), ]

  # Replace names in the Abundance column based on taxonomy rank
  df_filtered$Abundance <- gsub("^s__.*", "Species", df_filtered$Abundance)
  df_filtered$Abundance <- gsub("^g__.*", "Genus", df_filtered$Abundance)
  df_filtered$Abundance <- gsub("^f__.*", "Family", df_filtered$Abundance)
  df_filtered$Abundance <- gsub("^o__.*", "Order", df_filtered$Abundance)
  df_filtered$Abundance <- gsub("^c__.*", "Class", df_filtered$Abundance)
  df_filtered$Abundance <- gsub("^p__.*", "Phylum", df_filtered$Abundance)
  df_filtered$Abundance <- gsub("^k__.*", "Kingdom", df_filtered$Abundance)
  
  df_filtered$thresh <- threshold
  return(df_filtered)
}

# Processing zero threshold first to establish reference proportions
df1 <- species_abundance %>% select(Abundance, SAMEA110452924)
df_zero_threshold <- apply_threshold(df1, 0, "SAMEA110452924")
zero_summary <- df_zero_threshold %>%
  group_by(Abundance) %>%
  summarise(count_zero = n(), .groups = 'drop')

# Process other thresholds and calculate proportions relative to zero threshold
summary_list1 <- lapply(thresholds, function(threshold) {
  df_threshold <- apply_threshold(df1, threshold, "SAMEA110452924")
  df_summary <- df_threshold %>%
    group_by(thresh, Abundance) %>%
    summarise(count = n(), .groups = 'drop') %>%
    left_join(zero_summary, by = "Abundance") %>%
    mutate(prop = count / count_zero * 100)
  return(df_summary)
})

summary_df1 <- bind_rows(summary_list1)

abundance_matrix_new_prop <- summary_df1 %>%
  group_by(thresh, Abundance) %>%
  summarize(prop = mean(prop)) %>%
  pivot_wider(names_from = Abundance, values_from = prop) 

abundance_matrix_new_prop <- as.data.frame(abundance_matrix_new_prop)
rownames(abundance_matrix_new_prop) <- abundance_matrix_new_prop$thresh
abundance_matrix_new_prop$thresh <- NULL

abundance_matrix_new_prop <- abundance_matrix_new_prop %>%
  select(Kingdom, Phylum, Class, Order, Family, Genus, Species)

proportional_info_loss <- 100 - abundance_matrix_new_prop

new_df <- proportional_info_loss
for (i in 2:nrow(proportional_info_loss)) {
  new_df[i - 1, ] <- proportional_info_loss[i, ] - proportional_info_loss[i - 1, ]
}
new_df[nrow(proportional_info_loss), ] <- 100 - proportional_info_loss[nrow(proportional_info_loss), ]

new_df$Threshold <- rownames(new_df)
new_df <- new_df %>% select(Threshold, everything())

summary_df1 <- new_df %>%
  pivot_longer(cols = -Threshold, names_to = "Abundance", values_to = "prop")
summary_df1$Abundance <- factor(summary_df1$Abundance, levels = c("Kingdom", "Phylum", "Class", "Order", "Family", "Genus", "Species"))

# Processing zero threshold first to establish reference proportions
df2 <- species_abundance %>% select(Abundance, SAMEA110452918)
df_zero_threshold <- apply_threshold(df2, 0, "SAMEA110452918")
zero_summary <- df_zero_threshold %>%
  group_by(Abundance) %>%
  summarise(count_zero = n(), .groups = 'drop')

# Process other thresholds and calculate proportions relative to zero threshold
summary_list2 <- lapply(thresholds, function(threshold) {
  df_threshold <- apply_threshold(df2, threshold, "SAMEA110452918")
  df_summary <- df_threshold %>%
    group_by(thresh, Abundance) %>%
    summarise(count = n(), .groups = 'drop') %>%
    left_join(zero_summary, by = "Abundance") %>%
    mutate(prop = count / count_zero * 100)
  return(df_summary)
})

summary_df2 <- bind_rows(summary_list2)
abundance_matrix_new_prop <- summary_df2 %>%
  group_by(thresh, Abundance) %>%
  summarize(prop = mean(prop)) %>%
  pivot_wider(names_from = Abundance, values_from = prop) 

abundance_matrix_new_prop <- as.data.frame(abundance_matrix_new_prop)
rownames(abundance_matrix_new_prop) <- abundance_matrix_new_prop$thresh
abundance_matrix_new_prop$thresh <- NULL

abundance_matrix_new_prop <- abundance_matrix_new_prop %>%
  select(Kingdom, Phylum, Class, Order, Family, Genus, Species)
proportional_info_loss <- 100 - abundance_matrix_new_prop

new_df <- proportional_info_loss
for (i in 2:nrow(proportional_info_loss)) {
  new_df[i - 1, ] <- proportional_info_loss[i, ] - proportional_info_loss[i - 1, ]
}
new_df[nrow(proportional_info_loss), ] <- 100 - proportional_info_loss[nrow(proportional_info_loss), ]

new_df$Threshold <- rownames(new_df)
new_df <- new_df %>% select(Threshold, everything())

summary_df2 <- new_df %>%
  pivot_longer(cols = -Threshold, names_to = "Abundance", values_to = "prop")
summary_df2$Abundance <- factor(summary_df2$Abundance, levels = c("Kingdom", "Phylum", "Class", "Order", "Family", "Genus", "Species"))


### drop ref
#summary_df1 <- subset(summary_df1, thresh != "0")
#summary_df2 <- subset(summary_df2, thresh != "0" )

# Define colors
color <- c("#005f73","#0a9396","#80B1D3","#e9d8a6", "#ae2012")


# Create plot for SAMEA110452924
plot1 <- ggplot(summary_df1, aes(x = Abundance, y = prop, fill = as.factor(Threshold))) + 
  geom_bar(stat = "identity", position = position_fill(), colour = "black") + 
  theme(axis.text.x = element_text(angle = 90, size = 14, colour = "black", vjust = 0.5, hjust = 1, face = "bold"), 
        axis.title.y = element_text(size = 16, face = "bold"), 
        legend.title = element_text(size = 16, face = "bold"), 
        legend.text = element_text(size = 12, face = "bold", colour = "black"), 
        axis.text.y = element_text(colour = "black", size = 12, face = "bold"),
        plot.title = element_text(size = 20, face = "bold", hjust = 0.5, margin = margin(b = 10))) + 
  scale_y_continuous(expand = c(0, 0), labels = scales::percent) + 
  labs(title = "SAMEA110452924", x = "", y = "Proportion (%)", fill = "Threshold") +
  scale_fill_manual(values = color)

# Create plot for SAMEA110452918
plot2 <- ggplot(summary_df2, aes(x = Abundance, y = prop, fill = as.factor(Threshold))) + 
  geom_bar(stat = "identity", position = position_fill(), colour = "black") + 
  theme(axis.text.x = element_text(angle = 90, size = 14, colour = "black", vjust = 0.5, hjust = 1, face = "bold"), 
        axis.title.y = element_text(size = 16, face = "bold"), 
        legend.title = element_text(size = 16, face = "bold"), 
        legend.text = element_text(size = 12, face = "bold", colour = "black"), 
        axis.text.y = element_text(colour = "black", size = 12, face = "bold"),
        plot.title = element_text(size = 20, face = "bold", hjust = 0.5, margin = margin(b = 10))) + 
  scale_y_continuous(expand = c(0, 0), labels = scales::percent) + 
  labs(title = "SAMEA110452918", x = "", y = "Proportion (%)", fill = "Threshold (%)") +
  scale_fill_manual(values = color)

combined_plot <- plot1 + plot2

print(combined_plot)
```


```{r}
# Processing zero threshold first to establish reference proportions
df1 <- species_abundance %>% select(Abundance, SAMEA110452924)
df_zero_threshold <- apply_threshold(df1, 0, "SAMEA110452924")
zero_summary <- df_zero_threshold %>%
  group_by(Abundance) %>%
  summarise(count_zero = sum(SAMEA110452924), .groups = 'drop')

# Process other thresholds and calculate proportions relative to zero threshold
summary_list1 <- lapply(thresholds, function(threshold) {
  df_threshold <- apply_threshold(df1, threshold, "SAMEA110452924")
  df_summary <- df_threshold %>%
    group_by(thresh, Abundance) %>%
    summarise(count = sum(SAMEA110452924), .groups = 'drop') %>%
    left_join(zero_summary, by = "Abundance") %>%
    mutate(prop = count / count_zero * 100)
  return(df_summary)
})

summary_df1 <- bind_rows(summary_list1)
abundance_matrix_new_prop <- summary_df1 %>%
  group_by(thresh, Abundance) %>%
  summarize(prop = mean(prop)) %>%
  pivot_wider(names_from = Abundance, values_from = prop) 

abundance_matrix_new_prop <- as.data.frame(abundance_matrix_new_prop)
rownames(abundance_matrix_new_prop) <- abundance_matrix_new_prop$thresh
abundance_matrix_new_prop$thresh <- NULL

abundance_matrix_new_prop <- abundance_matrix_new_prop %>%
  select(Kingdom, Phylum, Class, Order, Family, Genus, Species)
proportional_info_loss <- 100 - abundance_matrix_new_prop

new_df <- proportional_info_loss
for (i in 2:nrow(proportional_info_loss)) {
  new_df[i - 1, ] <- proportional_info_loss[i, ] - proportional_info_loss[i - 1, ]
}
new_df[nrow(proportional_info_loss), ] <- 100 - proportional_info_loss[nrow(proportional_info_loss), ]

new_df$Threshold <- rownames(new_df)
new_df <- new_df %>% select(Threshold, everything())

summary_df1 <- new_df %>%
  pivot_longer(cols = -Threshold, names_to = "Abundance", values_to = "prop")
summary_df1$Abundance <- factor(summary_df1$Abundance, levels = c("Kingdom", "Phylum", "Class", "Order", "Family", "Genus", "Species"))

# Processing zero threshold first to establish reference proportions
df2 <- species_abundance %>% select(Abundance, SAMEA110452918)
df_zero_threshold <- apply_threshold(df2, 0, "SAMEA110452918")
zero_summary <- df_zero_threshold %>%
  group_by(Abundance) %>%
  summarise(count_zero = sum(SAMEA110452918), .groups = 'drop')

# Process other thresholds and calculate proportions relative to zero threshold
summary_list2 <- lapply(thresholds, function(threshold) {
  df_threshold <- apply_threshold(df2, threshold, "SAMEA110452918")
  df_summary <- df_threshold %>%
    group_by(thresh, Abundance) %>%
    summarise(count =sum(SAMEA110452918), .groups = 'drop') %>%
    left_join(zero_summary, by = "Abundance") %>%
    mutate(prop = count / count_zero * 100)
  return(df_summary)
})

# Process SAMEA110452918
summary_df2 <- bind_rows(summary_list2)
abundance_matrix_new_prop <- summary_df2 %>%
  group_by(thresh, Abundance) %>%
  summarize(prop = mean(prop)) %>%
  pivot_wider(names_from = Abundance, values_from = prop)

abundance_matrix_new_prop <- as.data.frame(abundance_matrix_new_prop)
rownames(abundance_matrix_new_prop) <- abundance_matrix_new_prop$thresh
abundance_matrix_new_prop$thresh <- NULL

abundance_matrix_new_prop <- abundance_matrix_new_prop %>%
  select(Kingdom, Phylum, Class, Order, Family, Genus, Species)
proportional_info_loss <- 100 - abundance_matrix_new_prop

new_df <- proportional_info_loss
for (i in 2:nrow(proportional_info_loss)) {
  new_df[i - 1, ] <- proportional_info_loss[i, ] - proportional_info_loss[i - 1, ]
}
new_df[nrow(proportional_info_loss), ] <- 100 - proportional_info_loss[nrow(proportional_info_loss), ]

new_df$Threshold <- rownames(new_df)
new_df <- new_df %>% select(Threshold, everything())

summary_df2 <- new_df %>%
  pivot_longer(cols = -Threshold, names_to = "Abundance", values_to = "prop")
summary_df2$Abundance <- factor(summary_df2$Abundance, levels = c("Kingdom", "Phylum", "Class", "Order", "Family", "Genus", "Species"))


### drop ref
#summary_df1 <- subset(summary_df1, thresh != "0")
#summary_df2 <- subset(summary_df2, thresh != "0" )

# Define colors
color <- c("#005f73","#0a9396","#80B1D3","#e9d8a6", "#ae2012")


# Create plot for SAMEA110452924
plot1 <- ggplot(summary_df1, aes(x = Abundance, y = prop, fill = as.factor(Threshold))) + 
  geom_bar(stat = "identity", position = position_fill(), colour = "black") + 
  theme(axis.text.x = element_text(angle = 90, size = 14, colour = "black", vjust = 0.5, hjust = 1, face = "bold"), 
        axis.title.y = element_text(size = 16, face = "bold"), 
        legend.title = element_text(size = 16, face = "bold"), 
        legend.text = element_text(size = 12, face = "bold", colour = "black"), 
        axis.text.y = element_text(colour = "black", size = 12, face = "bold"),
        plot.title = element_text(size = 20, face = "bold", hjust = 0.5, margin = margin(b = 10))) + 
  scale_y_continuous(expand = c(0, 0), labels = scales::percent) + 
  labs(title = "SAMEA110452924", x = "", y = "Proportion (%)", fill = "Threshold") +
  scale_fill_manual(values = color)

# Create plot for SAMEA110452918
plot2 <- ggplot(summary_df2, aes(x = Abundance, y = prop, fill = as.factor(Threshold))) + 
  geom_bar(stat = "identity", position = position_fill(), colour = "black") + 
  theme(axis.text.x = element_text(angle = 90, size = 14, colour = "black", vjust = 0.5, hjust = 1, face = "bold"), 
        axis.title.y = element_text(size = 16, face = "bold"), 
        legend.title = element_text(size = 16, face = "bold"), 
        legend.text = element_text(size = 12, face = "bold", colour = "black"), 
        axis.text.y = element_text(colour = "black", size = 12, face = "bold"),
        plot.title = element_text(size = 20, face = "bold", hjust = 0.5, margin = margin(b = 10))) + 
  scale_y_continuous(expand = c(0, 0), labels = scales::percent) + 
  labs(title = "SAMEA110452918", x = "", y = "Proportion (%)", fill = "Threshold (%)") +
  scale_fill_manual(values = color)

combined_plot <- plot1 + plot2

print(combined_plot)
```
